File: 2021/Day 10/Part 2/Day10_2.java
import java.util.*;

public class Day10_2 {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        Map<Character, Integer> autocompleteScores = new HashMap<>();
        autocompleteScores.put('(', 1);
        autocompleteScores.put('[', 2);
        autocompleteScores.put('{', 3);
        autocompleteScores.put('<', 4);

        List<Long> scores = new ArrayList<>();

        while (sc.hasNextLine()) {
            char[] line = sc.nextLine().toCharArray();
            long score = getAutocompleteScore(line, autocompleteScores);
            if (score != 0) scores.add(score);
        }

        Collections.sort(scores);

        System.out.println(scores.get(scores.size() / 2));
        sc.close();
    }

    private static long getAutocompleteScore(char[] line, Map<Character, Integer> autocompleteScores) {
        Stack<Character> stack = new Stack<>();

        for (char character : line) {
            switch (character) {
                case '(', '[', '{', '<' -> stack.push(character);
                case ')' -> { if (stack.pop() != '(') return 0; }
                case ']' -> { if (stack.pop() != '[') return 0; }
                case '}' -> { if (stack.pop() != '{') return 0; }
                case '>' -> { if (stack.pop() != '<') return 0; }
            }
        }

        long score = 0;

        while (!stack.isEmpty()) {
            score = score * 5 + autocompleteScores.get(stack.pop());
        }

        return score;
    }
}

File: 2021/Day 10/Part 2/day10_2.cpp
#include <algorithm>
#include <iostream>
#include <stack>
#include <unordered_map>
#include <vector>

using namespace std;

const unordered_map<char, char> pairs = { {')', '('}, {']', '['}, {'}', '{'}, {'>', '<'} };
const unordered_map<char, int> scores = { {'(', 1}, {'[', 2}, {'{', 3}, {'<', 4} };

long long getAutocompleteScore(string& line) {
    long long score = 0;
    stack<char> stack;

    for (char ch : line) {
        if (ch == '(' || ch == '[' || ch == '{' || ch == '<') {
            stack.push(ch);
        } else if (stack.top() == pairs.at(ch)) {
            stack.pop();
        } else {
            return score;
        }
    }

    while (!stack.empty()) {
        score = score * 5 + scores.at(stack.top());
        stack.pop();
    }

    return score;
}

int main() {
    vector<long long> scores;

    string line;
    while (cin >> line) {
        long long score = getAutocompleteScore(line);

        if (score != 0) {
            scores.push_back(score);
        }
    }

    sort(scores.begin(), scores.end());

    cout << scores[scores.size() / 2];

    return 0;
}
